osutil/sync_dir.go

```diff
@@ -0,0 +1,158 @@
+type FileState struct {
+	Content []byte
+	Mode    os.FileMode
+	UID     uint32
+	GID     uint32
+}
+
+// EnsureDirState ensures that directory content matches expectations.
+//
+// EnsureDirState enumerates all the files in the specified directory that
+// match the provided pattern (glob). Each enumerated file is checked to ensure
+// that the contents, permissions and ownership are what is desired. Unexpected
+// files are removed.  Missing files are created and differing files are
+// corrected.  Files not matching the pattern are ignored.
+//
+// The content map describes each of the files that are intended to exist in
+// the directory.  Map keys must be file names relative to the directory.
+// Sub-directories in the name are not allowed.
+//
+// The function stops at the first encountered error but reports all of the
+// changes performed so far. Information about the performed changes is
+// returned to the caller for any extra processing that might be required (e.g.
+// to run some helper program).
+//
+// The return value is: created, corrected, removed []string, err error
+func EnsureDirState(dir, glob string, content map[string]*FileState) ([]string, []string, []string, error) {
+	matches, err := filepath.Glob(path.Join(dir, glob))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	found := make(map[string]bool)
+	var created, corrected, removed []string
+	// Analyze files that inhabit the subset defined by our glob pattern.
+	for _, name := range matches {
+		baseName := path.Base(name)
+		// Remove files that should not be here.
+		var expected *FileState
+		var shouldBeHere bool
+		if expected, shouldBeHere = content[baseName]; !shouldBeHere {
+			if err := os.RemoveAll(name); err != nil {
+				return created, corrected, removed, err
+			}
+			removed = append(removed, baseName)
+			continue
+		}
+		var file *os.File
+		if file, err = os.OpenFile(name, os.O_RDWR, 0); err != nil {
+			return created, corrected, removed, err
+		}
+		defer file.Close()
+		var stat os.FileInfo
+		if stat, err = file.Stat(); err != nil {
+			return created, corrected, removed, err
+		}
+		found[baseName] = true
+		// Check that file has the right content
+		needsRewrite, err := fileNeedsRewrite(file, stat, expected)
+		if err != nil {
+			return created, corrected, removed, err
+		}
+		if needsRewrite {
+			if err := AtomicWriteFile(file.Name(), expected.Content, expected.Mode, 0); err != nil {
+				return created, corrected, removed, err
+			}
+			corrected = append(corrected, baseName)
+			// NOTE: rewriting files also fixes permissions so we can skip the last stage
+			continue
+		}
+		// Check that file has the right meta-data
+		changed := false
+		currentPerm := stat.Mode().Perm()
+		expectedPerm := expected.Mode.Perm()
+		if currentPerm != expectedPerm {
+			if err := file.Chmod(expectedPerm); err != nil {
+				return created, corrected, removed, err
+			}
+			changed = true
+		}
+		if st, ok := stat.Sys().(*syscall.Stat_t); ok {
```

Comment:
This has a suble bug that I've introduced  after switching to atomic writes. ``stat`` and ``file`` are not longer related to the newly-renamed file that was created by AtomicWrite()